File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsDefaultNamedPreferenceStorage.java
package co.infinum.princeofversions;

import android.content.Context;
import android.content.SharedPreferences;

import java.util.concurrent.Callable;

import javax.annotation.Nullable;

final class PrinceOfVersionsDefaultNamedPreferenceStorage implements Storage {

    private static final String KEY = "PrinceOfVersions_LastNotifiedUpdate";
    private static final String PREF_FILE_NAME = "co.infinum.princeofversions.PREF_FILE";

    private final SharedPreferences sp;

    PrinceOfVersionsDefaultNamedPreferenceStorage(final Context context) {
        sp = Lazy.create(SharedPreferences.class, new Callable<SharedPreferences>() {
            @Override
            public SharedPreferences call() {
                return context.getSharedPreferences(PREF_FILE_NAME, Context.MODE_PRIVATE);
            }
        });
    }

    @Nullable
    @Override
    public Integer lastNotifiedVersion(@Nullable Integer defaultValue) {
        try {
            String value = sp.getString(KEY, String.valueOf(defaultValue));
            if (value != null) {
                return Integer.valueOf(value);
            } else {
                return null;
            }
        } catch (NumberFormatException invalid) {
            return null;
        }
    }

    @Override
    public void rememberLastNotifiedVersion(@Nullable Integer version) {
        sp.edit().putString(KEY, String.valueOf(version)).apply();
    }
}

File: prince-of-versions/src/main/java/co/infinum/princeofversions/PrinceOfVersionsDefaultNamedPreferenceStorage.kt
package co.infinum.princeofversions

import android.content.Context
import android.content.SharedPreferences

internal class PrinceOfVersionsDefaultNamedPreferenceStorage(context: Context) : Storage {

    private companion object {
        private const val KEY = "PrinceOfVersions_LastNotifiedUpdate"
        private const val PRINCE_OF_VERSIONS_PREFERENCES = "co.infinum.princeofversions.PREF_FILE"
    }

    private val sharedPreferences: SharedPreferences by lazy {
        context.getSharedPreferences(PRINCE_OF_VERSIONS_PREFERENCES, Context.MODE_PRIVATE)
    }

    override fun lastNotifiedVersion(defaultValue: Int?): Int? {
        val value = sharedPreferences.getString(KEY, defaultValue.toString())
        return value?.toIntOrNull()
    }

    override fun rememberLastNotifiedVersion(version: Int?) {
        version?.let { sharedPreferences.edit().putString(KEY, version.toString()).apply() }
    }
}
